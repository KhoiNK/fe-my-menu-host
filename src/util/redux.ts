export const createActionName = (reducer: string, actionName: string) =>
  `@@${reducer.toUpperCase()}/${actionName}`;
