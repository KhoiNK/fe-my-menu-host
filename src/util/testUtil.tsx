import React, { ReactElement } from "react";
import { ThemeProvider } from "styled-components";
import theme from "../styles/theme";
import { render, RenderOptions, RenderResult } from "@testing-library/react";

const WithTheme = ({ children }: { children: React.ReactNode }) => {
  return <ThemeProvider theme={theme.main}>{children}</ThemeProvider>;
};

const renderTestWithTheme = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
): RenderResult => render(ui, { ...(options || {}), wrapper: WithTheme });

export * from "@testing-library/react";
export { renderTestWithTheme };
