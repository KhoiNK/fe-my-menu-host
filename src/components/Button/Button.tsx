import React, { HTMLAttributes } from "react";
import { ButtonStyled } from "./style";

export const enum BUTTON_TYPE_ENUM {
  primary = "primary",
  default = "default",
}

interface Props
  extends React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  > {
  text: string;
  buttonType?: keyof typeof BUTTON_TYPE_ENUM;
}

const Button: React.FC<Props> = ({ text, buttonType, ...rest }) => {
  return (
    <ButtonStyled buttonType={buttonType || "primary"} {...rest}>
      <b>{text}</b>
    </ButtonStyled>
  );
};

export default Button;
