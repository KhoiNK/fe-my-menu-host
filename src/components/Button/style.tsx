import styled from "styled-components";
import { toRem } from "../../styles/globalStyle";
import { Theme } from "../../styles/theme";
import { BUTTON_TYPE_ENUM } from "./Button";

const getButtonTypeBackground = (
  theme: Theme,
): {
  [key: keyof typeof BUTTON_TYPE_ENUM | string]: string;
} => ({
  primary: theme.colors.primary500,
  default: theme.colors.grey0,
});
const getButtonTypeColor = (
  theme: Theme,
): {
  [key: keyof typeof BUTTON_TYPE_ENUM | string]: string;
} => ({
  primary: theme.colors.grey0,
  default: theme.colors.primary1000,
});
const getButtonTypeBorder = (
  theme: Theme,
): {
  [key: keyof typeof BUTTON_TYPE_ENUM | string]: string;
} => ({
  primary: "none",
  default: `1px solid ${theme.colors.primary1000}`,
});

export const ButtonStyled = styled.button<{ buttonType: string }>`
  font-size: ${({ theme }) => theme.fontSize.bodyB2Regular};
  color: ${({ buttonType, theme }) => getButtonTypeColor(theme)[buttonType]};
  background-color: ${({ buttonType, theme }) =>
    getButtonTypeBackground(theme)[buttonType]};
  border-radius: ${({ theme }) => theme.borderRadius.button};
  text-align: center;
  border: ${({ buttonType, theme }) => getButtonTypeBorder(theme)[buttonType]};
  padding: ${toRem(16)};
  min-width: 360px;
  cursor: pointer;

  &:disabled {
    background-color: ${({ theme }) => theme.colors.grey200};
    color: ${({ theme }) => theme.colors.grey400};
  }
`;
