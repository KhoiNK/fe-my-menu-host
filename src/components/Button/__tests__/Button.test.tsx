import React from "react";
import { renderTestWithTheme } from "../../../util/testUtil";
import Button from "../Button";

describe("Buttons", () => {
  it("should render correctly", () => {
    const { container } = renderTestWithTheme(<Button text="Test button" />);
    expect(container).toMatchSnapshot();
  });
  it("should render default type", () => {
    const { container, queryByText } = renderTestWithTheme(
      <Button text="Test button" buttonType="default" />,
    );
    expect(container).toBeDefined();
    expect(queryByText("Test button")).toBeDefined();
  });
});
