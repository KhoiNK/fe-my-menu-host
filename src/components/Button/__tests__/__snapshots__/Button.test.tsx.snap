// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buttons should render correctly 1`] = `
<div>
  <button
    buttontype="primary"
    class="style__ButtonStyled-sc-103epc7-0 ekabZS"
  >
    <b>
      Test button
    </b>
  </button>
</div>
`;
